### PR TITLE
Remove stale lock table entries of invalid sessionid

### PR DESCRIPTION
**Pull request by asmithakarun**
Locks are cleaned up when there is a graceful logout /
session expiration due to timeout. But there are possibilities
of leaving stale lock entries in the BMC, if the session is
abruptly terminated.

This commit is to check for the existence of the session
whenever there is a lock conflict. If the session is not active,
remove the lock record considering that as a stale record and
return no error. If the session is active, return the lock
conflict to the client.

It fixes the following defect:
https://w3.rchland.ibm.com/projects/bestquest/?verb=view&id=SW538023&table=defect

Signed-off-by: Asmitha Karunanithi <asmitk01@in.ibm.com>
Change-Id: I1b97895cbdc4e602e00313bfd3b59e3b3445bd59

### CODE REVIEW

File: include/ibm/management_console_rest.hpp [hunk 1]
```diff
@@ -533,14 +533,32 @@ inline void
...
         auto var =
             std::get<std::pair<uint32_t, LockRequest>>(conflictStatus.second);
+        std::string sessionId = std::get<0>(var.second);
+        auto session =
+            persistent_data::SessionStore::getInstance().getSessionByUid(
+                sessionId);
+
+        if (session == nullptr)
+        {
+            BMCWEB_LOG_DEBUG << "Releasing lock of the session id(stale): "
```
Comment (sunharis) [inline include/ibm/management_console_rest.hpp:548]: Please change this to LOG_ERROR so that this gets printed in the journal

File: include/ibm/management_console_rest.hpp [hunk 2]
```diff
@@ -533,14 +533,32 @@ inline void
...
+                             << sessionId;
+            // clear the stale entry
+            crow::ibm_mc_lock::Lock::getInstance().releaseLock(sessionId);
+            return;
```
Comment (sunharis) [inline include/ibm/management_console_rest.hpp:552]: After releasing this stale lock - before returning the response you should continue processing the lock request - to actually set the lock in the lock table for this current command. 
By calling crow::ibm_mc_lock::Lock::getInstance().acquireLock(t) again, it will create a successful lock record for this new session

File: include/ibm/management_console_rest.hpp [hunk 3]
```diff
@@ -533,14 +533,32 @@ inline void
...
+        }
+
+        BMCWEB_LOG_DEBUG << "There is a conflict with the lock table";
```
Comment (sunharis) [inline include/ibm/management_console_rest.hpp:555]: Please convert this to ERROR log

**Review by sunharis — CHANGES_REQUESTED** [review]

File: include/ibm/management_console_rest.hpp (hunk 2 above)
Comment (asmithakarun) [reply include/ibm/management_console_rest.hpp]: done

File: include/ibm/management_console_rest.hpp (hunk 1 above)
Comment (asmithakarun) [reply include/ibm/management_console_rest.hpp]: Done

File: include/ibm/management_console_rest.hpp (hunk 3 above)
Comment (asmithakarun) [reply include/ibm/management_console_rest.hpp]: Done

File: include/ibm/management_console_rest.hpp [hunk 4]
```diff
@@ -533,14 +533,45 @@ inline void
...
+            varAcquireLock =
+                crow::ibm_mc_lock::Lock::getInstance().acquireLock(t);
+            conflictStatus =
+                std::get<crow::ibm_mc_lock::Rc>(varAcquireLock.second);
+            if (!conflictStatus.first)
+            {
+                auto var = std::get<uint32_t>(conflictStatus.second);
+                nlohmann::json returnJson;
+                returnJson["id"] = var;
+                asyncResp->res.jsonValue["TransactionID"] = var;
```
Comment (sunharis) [inline include/ibm/management_console_rest.hpp:563]: asyncResp->res.result(boost::beast::http::status::ok); should be set

**Review by sunharis — CHANGES_REQUESTED** [review]

File: include/ibm/management_console_rest.hpp (hunk 4 above)
Comment (asmithakarun) [reply include/ibm/management_console_rest.hpp:565]: Done

File: include/ibm/management_console_rest.hpp [hunk 5]
```diff
@@ -533,14 +533,47 @@ inline void
...
+                crow::ibm_mc_lock::Lock::getInstance().acquireLock(t);
+            conflictStatus =
+                std::get<crow::ibm_mc_lock::Rc>(varAcquireLock.second);
+            if (!conflictStatus.first)
+            {
+                asyncResp->res.result(boost::beast::http::status::ok);
+
+                auto var = std::get<uint32_t>(conflictStatus.second);
+                nlohmann::json returnJson;
+                returnJson["id"] = var;
```
Comment (manojkiraneda) [inline include/ibm/management_console_rest.hpp:564]: why was this needed ?

**Review by sunharis — APPROVED** [review]

File: include/ibm/management_console_rest.hpp [hunk 6]
```diff
@@ -533,14 +533,47 @@ inline void
...
-        asyncResp->res.result(boost::beast::http::status::conflict);
+
+        // Check if the session is active. If active, send back the lock
+        // conflict, else clear the stale entry in the lock table.
         auto var =
             std::get<std::pair<uint32_t, LockRequest>>(conflictStatus.second);
+        std::string sessionId = std::get<0>(var.second);
+        auto session =
+            persistent_data::SessionStore::getInstance().getSessionByUid(
+                sessionId);
```
Comment (manojkiraneda) [inline include/ibm/management_console_rest.hpp:544]: do we wanna do this for every conflicting request ?

Comment (manojkiraneda) [reply include/ibm/management_console_rest.hpp:544]: cant we do it just once in the constructor, so we remove stale entries in one shot when bmcweb comes up ? rather than checking them every time ?

**Review by manojkiraneda — COMMENTED** [review]
How was this tested ? can we get some testing via redfish or via unit tests ?

File: include/ibm/management_console_rest.hpp (hunk 6 above)
Comment (sunharis) [reply include/ibm/management_console_rest.hpp:544]: With the several defects data - what we saw is the locks are going stale while the bmcweb/bmc is running. Somehow the session from HMC is getting vanished with-out a proper cleanup path hitting at bmc. This is leaving stale locks. 
So we need a way to check if the conflicting lock is associated with any stale session OR and active session

File: include/ibm/management_console_rest.hpp (hunk 5 above)
Comment (asmithakarun) [reply include/ibm/management_console_rest.hpp:564]: so the following actually would have been reported as a lock conflict earlier
Now what it does is, removes the invalid session stale entry then creates a lock record for the current session and returns the json


curl -k -H "X-Auth-Token:$bmc_token" -X POST - "Content-type: application/json" -d '{ "Request" :[ { "LockType":"Write", "SegmentFlags": [ {"LockFlag":"DontLock","SegmentLength":1}, {"LockFlag":"DontLock","SegmentLength":2}, {"LockFlag":"LockAll","SegmentLength":1}], "ResourceID": 288241371267989504 } ]}' https://${bmc}/ibm/v1/HMC/LockService/Actions/LockService.AcquireLock
{
  "TransactionID": 2
}

Comment (manojkiraneda) [reply include/ibm/management_console_rest.hpp:564]: no i was asking ,
```  nlohmann::json returnJson;
     returnJson["id"] = var;
```
why was this needed ?

File: include/ibm/management_console_rest.hpp (hunk 6 above)
Comment (manojkiraneda) [reply include/ibm/management_console_rest.hpp:544]: hmm, so this is ideally a bug in bmcweb where the sessions are removed in a path that we don't know (not logout/timeout) - this is essentially a workaround of that bug.
